Change light.toggle service call to invoke LightEntity.async_toggle

ams2990 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
`LightEntity` overrides `ToggleEntity`, but the `light.toggle` service call never invokes `LightEntity.async_toggle`. This makes it impossible to write an integration that wants to handle toggle differently than a simple "if on, turn off; otherwise, turn on".

This PR changes that, so that `light.toggle` now invokes `LightEntity.async_toggle`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/2986
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
    - No: existing tests adequately cover the change in functionality. I know, because they caught a bug in my initial implementation :)
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.
    - n/a, no generated code

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

thecode commented:
> LightEntity overrides ToggleEntity, but the light.toggle service call never invokes LightEntity.async_toggle. This makes it impossible to write a custom integration that wants to handle toggle differently than a simple "if on, turn off; otherwise, turn on".

I am trying to understand why you can't override this method (`async_handle_toggle_service`) in your custom integration and handle this differently.
This method is not marked as `final`, meaning it was considered that it might be overridden in a sub-class.

ams2990 commented:
> I am trying to understand why you can't override this method (`async_handle_toggle_service`) in your custom integration and handle this differently. This method is not marked as `final`, meaning it was considered that it might be overridden in a sub-class.

Real answer: I wasn't aware that you could override a service in a custom integration -- I'm a relative novice to Home Assistant.

That said, it seems unnecessary to do so? LightEntity already nominally provides the correct extension point (`async_toggle`), and it's arguably a bug that overriding it doesn't work. The toggle service not calling toggle on the class is a surprising implementation.

emontnemery commented:
> I am trying to understand why you can't override this method (async_handle_toggle_service) in your custom integration and handle this differently.

It can't be overridden because it's not a class method, it's a regular function

emontnemery commented:
@ams2990 It sounds like what you really want is a way to set light parameters without changing the state, i.e. a new service separate from turn_on / turn_off or toggle?

emontnemery reviewed (approved):
I discussed the proposal with some others and we agree there's no need to block integrations from implementing their own toggle handlers

emontnemery commented:
@ams2990 wouldn't a custom service for your integration be easier for users to understand than hacking the toggle service?

ams2990 commented on homeassistant/components/light/__init__.py:
```diff
@@ -665,10 +672,8 @@ async def async_handle_toggle_service(
-        else:
-            await async_handle_light_on_service(light, call)
+        params: dict[str, Any] = _process_turn_on_params(light, call)
+        await light.async_toggle(**params)
```
I marked the PR as a breaking change for this reason, but I think this is a theoretical breakage. I can't immediately identify what problems this is likely to cause.

For instance, I think the following situation is required:
1) Someone is currently passing params to toggle that only async_turn_on accepts
2) They are using a custom integration whose async_turn_off misbehaves when receiving these.

I audited the HA repo to see if there were any integrations meeting criteria #2, but I couldn't find any. That doesn't mean there aren't out-of-tree integrations that won't be broken, however.

arturpragacz replied:
That's not the only change. You're also missing a call to `filter_turn_on_params` now.

Both of those changes are not acceptable.

You should implement it in a way that if async_toggle is not overridden in the integration, then nothing changes compared to the status quo.

ams2990 committed:
Refactor param manipulation in on/off service call

This commit is intended to be a no-op, to allow for functional changes
in subsequent PRs.

ams2990 committed:
light.toggle now invokes LightEntity.async_toggle

ams2990 committed:
Delete stale comment about white value

This comment was added in 5f2326f. The code it was attached to was
deleted in 60c8d95 but the comment was not removed.

arturpragacz reviewed (changes requested):
Overall looks good, left one comment.

The PR also needs to be updated to match the current `dev`.

ams2990 committed:
Addressed code review comments

Co-authored-by: Artur Pragacz <49985303+arturpragacz@users.noreply.github.com>

ams2990 commented:
Done.

ams2990 commented:
Looks like a bad merge. Looking...

ams2990 committed:
Fix bad manual merge

ams2990 committed:
Add method docstrings

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -272,6 +272,16 @@ def filter_turn_off_params(
+def process_turn_off_params(
+    hass: HomeAssistant, light: LightEntity, params: dict[str, Any]
+) -> dict[str, Any]:
+    """Translate light.turn_off service call parameters to light-supported attributes."""
```
```suggestion
    """Process light turn off params."""
```

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -306,7 +316,193 @@ def filter_turn_on_params(light: LightEntity, params: dict[str, Any]) -> dict[st
+def process_turn_on_params(  # noqa: C901
+    hass: HomeAssistant, light: LightEntity, params: dict[str, Any]
+) -> dict[str, Any]:
+    """Translate light.turn_on service call parameters to light-supported attributes."""
```
```suggestion
    """Process light turn on params."""
```

arturpragacz committed:
Adjust docstrings

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -1163,3 +1168,15 @@ def __should_report_light_issue(self) -> bool:
+
+    @override
+    async def async_toggle(self, **kwargs: Any) -> None:
+        params = kwargs
```
```suggestion
```

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -534,21 +544,16 @@ async def async_handle_light_off_service(
-        else:
-            await async_handle_light_on_service(light, call)
+        params = dict(call.data["params"])
+        await light.async_toggle(**params)
```
```suggestion
        await light.async_toggle(**call.data["params"])
```

arturpragacz reviewed (commented):
I've just realised that the dict copying is not quite right.

arturpragacz committed:
Fix dict copying

arturpragacz committed:
Fix dict copying 2

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -306,7 +318,195 @@ def filter_turn_on_params(light: LightEntity, params: dict[str, Any]) -> dict[st
+) -> dict[str, Any]:
+    """Process light turn on params."""
+    params = dict(params)
+    
```
```suggestion

```

arturpragacz committed:
Fix ruff

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -1163,3 +1171,14 @@ def __should_report_light_issue(self) -> bool:
         return self.platform.platform_name != "philips_js"
+
+    @override
+    async def async_toggle(self, **kwargs: Any) -> None:
```
```suggestion
    async def async_toggle(self, **kwargs: Any) -> None:
    """Toggle the entity."""
```

arturpragacz committed:
Add docstring

arturpragacz commented on homeassistant/components/light/__init__.py:
```diff
@@ -1163,3 +1171,15 @@ def __should_report_light_issue(self) -> bool:
+
+    @override
+    async def async_toggle(self, **kwargs: Any) -> None:
+    """Toggle the entity."""
```
```suggestion
        """Toggle the entity."""
```

arturpragacz committed:
Fix docstring

arturpragacz commented:
To speed things up a little, I fixed the last few details.

arturpragacz reviewed (approved):
Please adjust the description so it correctly reflects the current PR state, it should no longer be listed as breaking.

Also please update [the developer docs](https://developers.home-assistant.io/docs/core/entity/light/#methods) and link it.

ams2990 committed:
Merge branch 'dev' into toggle

ams2990 commented:
Documentation added in https://github.com/home-assistant/developers.home-assistant/pull/2986.

ams2990 commented on homeassistant/components/light/__init__.py:
```diff
@@ -510,21 +524,15 @@ async def async_handle_light_off_service(
-            await async_handle_light_off_service(light, call)
-        else:
-            await async_handle_light_on_service(light, call)
+        await light.async_toggle(**call.data["params"])
```
I agree that would be a reasonable test to add. That said, it doesn't seem to be an existing pattern in this file. That is, there is no test ensuring that `light.turn_on` calls `async_turn_on`, or `light.turn_off` calls `async_turn_off`. If maintainers so desire, I can add tests for all three of these.

emontnemery committed:
Merge branch 'dev' into toggle
